### PR TITLE
SONAR-7553 WS api/ce/activity_status provide CE activity related metrics

### PR DESCRIPTION
**Pull request by teryk**
@julienlancelot please review only my commit about the new WS

### CODE REVIEW

File: server/sonar-web/src/main/js/apps/background-tasks/store/actions.js [hunk 1]
```diff
@@ -145,27 +144,23 @@ function fetchTasks (filters) {
...
+      const [activity] = responses;
       const tasks = activity.tasks;
-      const total = activity.paging.total;
 
-      dispatch(receiveTasks(tasks, total));
+      dispatch(receiveTasks(tasks));
 
-      const pendingCount = pendingActivity.paging.total;
-      const failingCount = failingActivity.paging.total;
+      // FIXME request real numbers
```
Comment (SonarTech) [inline server/sonar-web/src/main/js/apps/background-tasks/store/actions.js]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://dory.sonarsource.com/coding_rules#rule_key=javascript%3AS1134)

File: server/sonar-server/src/main/java/org/sonar/server/ce/ws/ActivityStatusAction.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+  private void checkPermissions(@Nullable String componentUuid) {
+    if (componentUuid == null) {
+      userSession.checkPermission(GlobalPermissions.SYSTEM_ADMIN);
+    } else {
+      userSession.checkComponentUuidPermission(UserRole.ADMIN, componentUuid);
+    }
+  }
+
+  private boolean hasComponentInRequest(ActivityStatusWsRequest request) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/ce/ws/ActivityStatusAction.java]: MAJOR Make "hasComponentInRequest" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

File: server/sonar-server/src/main/java/org/sonar/server/ce/ws/ActivityStatusAction.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    return request.getComponentId() != null || request.getComponentKey() != null;
+  }
+
+  private ActivityStatusWsRequest toWsRequest(Request request) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/ce/ws/ActivityStatusAction.java]: MAJOR Make "toWsRequest" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* MAJOR 3 major

Watch the comments in this conversation to review them.

File: server/sonar-web/src/main/webapp/WEB-INF/db/migrate/1125_add_index_ce_activity_islast_status.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program; if not, write to the Free Software Foundation,
+# Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+
+#
+# SonarQube 5.5
+# SONAR-7187 & SONAR-7553
+#
+class AddIndexCeActivityIslastStatus < ActiveRecord::Migration
```
Comment (julienlancelot) [inline server/sonar-web/src/main/webapp/WEB-INF/db/migrate/1125_add_index_ce_activity_islast_status.rb]: The migration should have been added in a separate commit

File: sonar-db/src/main/java/org/sonar/db/ce/CeActivityMapper.java [hunk 5]
```diff
@@ -37,7 +38,7 @@
 
   List<CeActivityDto> selectOlderThan(@Param("beforeDate") long beforeDate);
 
-  int countByQuery(@Param("query") CeTaskQuery query);
+  int countLastByStatusAndComponentUuid(@Param("status") CeActivityDto.Status status, @Nullable @Param("componentUuid") String componentUuid);
```
Comment (julienlancelot) [inline sonar-db/src/main/java/org/sonar/db/ce/CeActivityMapper.java]: I would have also put the new SQL query into a separate commit

File: sonar-db/src/main/resources/org/sonar/db/ce/CeQueueMapper.xml [hunk 6]
```diff
@@ -29,8 +29,13 @@
     where cq.uuid=#{uuid}
   </select>
 
-  <select id="countByStatus" parameterType="org.sonar.db.ce.CeQueueDto$Status" resultType="int">
-    select count(1) from ce_queue where status=#{status}
+  <select id="countByStatusAndComponentUuid" parameterType="map" resultType="int">
+    select count(1)
+    from ce_queue
+    where status=#{status}
```
Comment (julienlancelot) [inline sonar-db/src/main/resources/org/sonar/db/ce/CeQueueMapper.xml]: You should ask the Atlas team if we should not add an index on CE_QUEUE.STATUS, but I really think we should !

File: server/sonar-server/src/main/java/org/sonar/server/ce/ws/ActivityStatusAction.java [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+      component = componentFinder.getByUuidOrKey(dbSession, request.getComponentId(), request.getComponentKey(), COMPONENT_ID_AND_KEY);
+    }
+    return Optional.fromNullable(component);
+  }
+
+  private void checkPermissions(@Nullable String componentUuid) {
+    if (componentUuid == null) {
+      userSession.checkPermission(GlobalPermissions.SYSTEM_ADMIN);
+    } else {
+      userSession.checkComponentUuidPermission(UserRole.ADMIN, componentUuid);
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/ce/ws/ActivityStatusAction.java]: This method is enough, as it's first searching for global permission

File: server/sonar-server/src/main/java/org/sonar/server/ce/ws/ActivityStatusAction.java [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    ComponentDto component = null;
+    if (hasComponentInRequest(request)) {
+      component = componentFinder.getByUuidOrKey(dbSession, request.getComponentId(), request.getComponentKey(), COMPONENT_ID_AND_KEY);
+    }
+    return Optional.fromNullable(component);
+  }
+
+  private void checkPermissions(@Nullable String componentUuid) {
+    if (componentUuid == null) {
+      userSession.checkPermission(GlobalPermissions.SYSTEM_ADMIN);
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/ce/ws/ActivityStatusAction.java]: Useless (see following comment)

File: server/sonar-server/src/main/java/org/sonar/server/ce/ws/ActivityStatusAction.java [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    writeProtobuf(activityStatusResponse, request, response);
+  }
+
+  private ActivityStatusWsResponse doHandle(ActivityStatusWsRequest request) {
+    DbSession dbSession = dbClient.openSession(false);
+    try {
+      Optional<ComponentDto> component = searchComponent(dbSession, request);
+      String componentUuid = component.isPresent() ? component.get().uuid() : null;
+      checkPermissions(componentUuid);
+      int pendingCount = dbClient.ceQueueDao().countByStatusAndComponentUuid(dbSession, CeQueueDto.Status.PENDING, componentUuid);
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/ce/ws/ActivityStatusAction.java]: Please use static imports for Statuses

**Comment by julienlancelot** [comment]
Split commits, add missing index and LGTM !

File: server/sonar-server/src/main/java/org/sonar/server/ce/ws/ActivityStatusAction.java (hunk 9 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/ce/ws/ActivityStatusAction.java]: I wanted to differentiate status from queue and from activity because it can be confusing

File: server/sonar-server/src/main/java/org/sonar/server/ce/ws/ActivityStatusAction.java (hunk 7 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/ce/ws/ActivityStatusAction.java]: as discussed, it should stay as is as `userSession.checkComponentUuidPermission(String,String)` does not accept a @Nullable component uuid and rightly so

File: server/sonar-server/src/main/java/org/sonar/server/ce/ws/ActivityStatusAction.java (hunk 9 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/ce/ws/ActivityStatusAction.java]: Ok, I haven't noticed that, no pb then

File: server/sonar-server/src/main/java/org/sonar/server/ce/ws/ActivityStatusAction.java (hunk 2 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/ce/ws/ActivityStatusAction.java]: k

File: server/sonar-server/src/main/java/org/sonar/server/ce/ws/ActivityStatusAction.java (hunk 7 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/ce/ws/ActivityStatusAction.java]: indeed

File: server/sonar-server/src/main/java/org/sonar/server/ce/ws/ActivityStatusAction.java (hunk 3 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/ce/ws/ActivityStatusAction.java]: ok

File: server/sonar-web/src/main/webapp/WEB-INF/db/migrate/1125_add_index_ce_activity_islast_status.rb (hunk 4 above)
Comment (teryk) [inline server/sonar-web/src/main/webapp/WEB-INF/db/migrate/1125_add_index_ce_activity_islast_status.rb]: I agree

File: sonar-db/src/main/java/org/sonar/db/ce/CeActivityMapper.java (hunk 5 above)
Comment (teryk) [inline sonar-db/src/main/java/org/sonar/db/ce/CeActivityMapper.java]: debatable, but ok

File: sonar-db/src/main/resources/org/sonar/db/ce/CeQueueMapper.xml (hunk 6 above)
Comment (teryk) [inline sonar-db/src/main/resources/org/sonar/db/ce/CeQueueMapper.xml]: indeed, I will add an index on CE_QUEUE_STATUS

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 2 major

Watch the comments in this conversation to review them.

File: sonar-db/src/main/resources/org/sonar/db/ce/CeQueueMapper.xml (hunk 6 above)
Comment (teryk) [inline sonar-db/src/main/resources/org/sonar/db/ce/CeQueueMapper.xml]: after discussion, I will NOT add an index on CE_QUEUE_STATUS for several reasons:

- for an index to be efficient, the filter must represent less than 10% of the table (our nominal case filter on 'PENDING' status)
- an index slows insert/update/delete operations and it's critical to have fast inserts on the CE_QUEUE table
- in general, we don't add an index for small table. I know the table could potentially be huge, but that's a sign of a problem. The table size should tend towards zero. Anyway, as stated, adding an index would help in this case

**Comment by teryk** [comment]
in the end I removed the useless call to count on `CE_QUEUE` table when calling WS api/ce/activity

**Comment by teryk** [comment]
ticket created for 5.6 to optimize call to CE_QUEUE and CE_ACTIVITY only when relevant (related to statuses given in parameter and the number of elements returned by the queue)
